cluster/mesos/cluster.go

```diff
@@ -0,0 +1,453 @@
+		d, err := time.ParseDuration(offerTimeout)
+		if err != nil {
+			return nil, err
+		}
+		cluster.offerTimeout = d
+	}
+
+	driver, err := mesosscheduler.NewMesosSchedulerDriver(driverConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	cluster.driver = driver
+
+	status, err := driver.Start()
+	if err != nil {
+		log.Debugf("Mesos driver started, status/err %v: %v", status, err)
+		return nil, err
+	}
+	log.Debugf("Mesos driver started, status %v", status)
+
+	return cluster, nil
+}
+
+// RegisterEventHandler registers an event handler.
+func (c *Cluster) RegisterEventHandler(h cluster.EventHandler) error {
+	if c.eventHandler != nil {
+		return errors.New("event handler already set")
+	}
+	c.eventHandler = h
+	return nil
+}
+
+// CreateContainer for container creation in Mesos task
+func (c *Cluster) CreateContainer(config *cluster.ContainerConfig, name string) (*cluster.Container, error) {
+	task, err := newTask(c, config, name)
+	if err != nil {
+		return nil, err
+	}
+
+	go c.pendingTasks.Add(task)
+
+	select {
+	case container := <-task.container:
+		return formatContainer(container), nil
+	case err := <-task.error:
+		return nil, err
+	case <-time.After(taskCreationTimeout):
+		c.pendingTasks.Remove(task)
+		return nil, strategy.ErrNoResourcesAvailable
+	}
+}
+
+// RemoveContainer to remove containers on mesos cluster
+func (c *Cluster) RemoveContainer(container *cluster.Container, force bool) error {
+	c.scheduler.Lock()
+	defer c.scheduler.Unlock()
+
+	return container.Engine.RemoveContainer(container, force)
+}
+
+// Images returns all the images in the cluster.
+func (c *Cluster) Images() []*cluster.Image {
+	c.RLock()
+	defer c.RUnlock()
+
+	out := []*cluster.Image{}
+	for _, s := range c.slaves {
+		out = append(out, s.engine.Images()...)
+	}
+
+	return out
+}
+
+// Image returns an image with IdOrName in the cluster
+func (c *Cluster) Image(IDOrName string) *cluster.Image {
+	// Abort immediately if the name is empty.
+	if len(IDOrName) == 0 {
+		return nil
+	}
+
+	c.RLock()
+	defer c.RUnlock()
+
+	for _, s := range c.slaves {
+		if image := s.engine.Image(IDOrName); image != nil {
+			return image
+		}
+	}
+
+	return nil
+}
+
+// RemoveImages removes images from the cluster
+func (c *Cluster) RemoveImages(name string) ([]*dockerclient.ImageDelete, error) {
+	return nil, errNotSupported
+}
+
+// Containers returns all the containers in the cluster.
+func (c *Cluster) Containers() cluster.Containers {
+	c.RLock()
+	defer c.RUnlock()
+
+	out := cluster.Containers{}
+	for _, s := range c.slaves {
+		for _, container := range s.engine.Containers() {
+			if name := container.Config.Labels[cluster.SwarmLabelNamespace+".mesos.name"]; name != "" && container.Names[0] != "/"+name {
+				container.Names = append([]string{"/" + name}, container.Names...)
+			}
+			out = append(out, container)
+		}
+	}
+
+	return out
+}
+
+func formatContainer(container *cluster.Container) *cluster.Container {
+	if container != nil {
```

Comment:
nit: `if container == nil { return nil }` instead of having lots of code within a nested if